### PR TITLE
feat(LOC-503): add MagicSync Viewer component

### PR DESCRIPTION
**Pull request by flycrum**
**Audience:** Users | Engineers | Third-Party Developers

**Summary:** The Magic Sync Viewer will give Local Team users the power to selectively choose which files they pull or push to Flywheel staging or production environments. In addition, users will be able to whitelist files to always ignore.

**Technical:** This builds on the VirtualList and Virtual table components. All data is mocked up to reflect the file structure of a real WordPress site.

**Screenshots:**
image
image
image
image

### CODE REVIEW

File: src/components/MagicSyncViewer/MagicSyncViewer.tsx [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+		{key: 'filename', value: 'Filename', className: styles.MagicSyncViewer_Column_Filename},
+		{key: 'localDate', value: 'Local', className: styles.MagicSyncViewer_Column_LocalDate},
+		{key: 'gtlt', value: '', className: styles.MagicSyncViewer_Column_DateIcons},
+		{key: 'flywheelDate', value: 'Flywheel (Staging)', className: styles.MagicSyncViewer_Column_FlywheelDate},
+	];
+
+	constructor (props: IMagicSyncViewerProps) {
+		super(props);
+
+		this.state ={
```
Comment (flycrum) [inline src/components/MagicSyncViewer/MagicSyncViewer.tsx]: A lot of this data will be clearer and passed around when MobX is introduced.

File: src/components/VirtualList/VirtualList.tsx [hunk 2]
```diff
@@ -13,7 +13,7 @@ export interface IVirtualListProps extends IReactComponentProps {
 	 */
 	containerRenderer? (children: React.ReactNode, customRendererStyles: {minHeight: number}): ReactElement<any> | null;
 	/** list data passed to the list item renderer */
-	data: any[];
+	data: any[] | undefined | null;
```
Comment (flycrum) [inline src/components/VirtualList/VirtualList.tsx]: This was missed initially.

File: src/components/MagicSyncViewer/helpers/MagicSyncViewerCellRenderers.tsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import { MagicSyncViewerCellRendererFilename } from './MagicSyncViewerCellRendererFilename';
+import FileGtSVG from '../../../svg/magic-sync-file-gt';
+import FileLtSVG from '../../../svg/magic-sync-file-lt';
+import FileAddedSVG from '../../../svg/magic-sync-file-added';
+import FileRemovedSVG from '../../../svg/magic-sync-file-removed';
+import LockLockedSVG from '../../../svg/magic-sync-lock-locked';
+import LockUnlockedSVG from '../../../svg/magic-sync-lock-unlocked';
+import { IVirtualTableCellRendererDataArgs, VirtualTableCellRenderer } from '../../VirtualTable/VirtualTable';
+
+export class MagicSyncViewerCellRenderers {
```
Comment (claygriffiths) [inline src/components/MagicSyncViewer/helpers/MagicSyncViewerCellRenderers.tsx]: I dig how these renderers were implemented!

File: src/components/MagicSyncViewer/components/MagicSyncViewerStactionbar.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import * as styles from './MagicSyncViewerStactionbar.scss';
+import IReactComponentProps from '../../../common/structures/IReactComponentProps';
+import FlySelect from '../../FlySelect';
+import { ConnectDirectionType } from '../helpers/magicSyncViewMock';
+
+interface IMagicSyncViewerHeaderProps extends IReactComponentProps {
+	connectDirection: ConnectDirectionType;
+}
+
+export const MagicSyncViewerStactionbar = (props: IMagicSyncViewerHeaderProps) => (
```
Comment (claygriffiths) [inline src/components/MagicSyncViewer/components/MagicSyncViewerStactionbar.tsx]: Naming is a bit confusing here. Might be worth sticking to one half of the portmanteau 😃

Comment (flycrum) [reply src/components/MagicSyncViewer/components/MagicSyncViewerStactionbar.tsx]: Ha, absolutely! I'll change it to `statusbar`. Good call.

**Pull request merged**